toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,91 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(bool first_name_component,
+                                              llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
+    if (parent.inst_id == SemIR::InstId::PackageNamespace) {
+      break;
+    }
+    if (!first_name_component) {
+      os << '.';
+    }
+    first_name_component = false;
+    CARBON_KIND_SWITCH(sem_ir.insts().Get(parent.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        const auto& impl = sem_ir.impls().Get(impl_decl.impl_id);
+        if (auto opt_class_self =
+                types.TryGetAs<SemIR::ClassType>(impl.self_id)) {
+          MangleInverseQualifiedNameScope(
+              true, os,
+              sem_ir.classes().Get(opt_class_self->class_id).scope_id);
+        } else {
+          auto builtin_self = types.GetAs<SemIR::BuiltinInst>(impl.self_id);
+          os << builtin_self.builtin_inst_kind.label();
+        }
+        os << ':';
+        auto opt_interface_constraint =
+            types.GetAs<SemIR::InterfaceType>(impl.constraint_id);
+        os << names.GetFormatted(sem_ir.interfaces()
+                                     .Get(opt_interface_constraint.interface_id)
+                                     .name_id);
+        return;
+      }
+      case CARBON_KIND(SemIR::ClassDecl class_decl): {
+        auto& class_node = sem_ir.classes().Get(class_decl.class_id);
+        os << names.GetFormatted(class_node.name_id);
```

Comment:
I think we typically append "_info" for `class` disambiguation (similarly `ns` for namespace). `_node` has a different meaning, since we do have nodes in parse.

```suggestion
        auto& class_info = sem_ir.classes().Get(class_decl.class_id);
        os << names.GetFormatted(class_info.name_id);
```